modin/engines/base/frame/partition_manager.py

```diff
@@ -260,6 +260,56 @@ def broadcast_axis_partitions(
         # the structure to the correct order.
         return result_blocks.T if not axis else result_blocks
 
+    @classmethod
+    def map_axis_partitions_to_another(
+        cls,
+        axis,
+        apply_func,
+        left,
+        right,
+    ):
+        """
+        Applies `apply_func` to every partition.
+
+        Parameters
+        ----------
+            axis : 0 or 1
+                The `self` axis to perform the map across (0 - index, 1 - columns).
+                The `other` axis is `axis ^ 1`.
+            partitions : NumPy array
+                The partitions of Modin Frame.
+            apply_func : callable
+                The function to apply.
+
+        Returns
+        -------
+        NumPy array
+            An array of new partitions for Modin Frame.
+
+        Notes
+        -----
+        This method should be used in the case that `apply_func` relies on
+        some global information about the axis.
+        """
+        preprocessed_map_func = cls.preprocess_func(apply_func)
+        left_partitions = cls.axis_partition(left, axis)
+        right_partitions = cls.axis_partition(right, axis ^ 1)
```

Comment:
Am I right that the only difference here from [`broadcast_axis_partitions`](https://github.com/modin-project/modin/blob/2b0b755753988cefbab38a71db973a26cbe581f4/modin/engines/base/frame/partition_manager.py#L210) is that we're specifying different axes for left and right partition? If it is, maybe we can make some common function for `broadcast_axis_partitions` and `map_axis_partitions_to_another` that will obtain two axis parameters: for the left and for the right partitions?

Reply:
No, there are some differences between these functions regarding to processing data.
I'd like to extend the logiс for `map_axis_partitions_to_another` as follows:
1) Add a flag (something like that - `mat_mul_behavior`).
2) In depend of that flag either process data as it is now or apply left partitions to right partitions keeping partitioning. It would look something like this:
```python
result_blocks = []
for l_part in left_partitions:
      for r_part in right_partitions:
          result_blocks.append(
              l_part.apply(
                    preprocessed_map_func,
                    num_splits=cls._compute_num_partitions(),
                    other_axis_partition=r_part,
               )
        )
result_blocks = np.array(result_blocks)
```
It would be useful for other operations (for example, `merge` with `inner` type of join). I was doing some measurements and my guessing was `merge` will be faster than as it is now, but it turned out `merge` was significantly slow. I think it is related to multiple transfer data (objects of `...Partition` classes which are responsible for keeping data). @devin-petersohn , what do you think about that?
So, I wouldn't join functions `broadcast_axis_partitions` and `map_axis_partitions_to_another`.

Reply:
I don't follow why a new operator is needed here.

Reply:
Here we also could apply each row partition of `self` on each column partition of `other` in parallel. As I thought, it could improve `merge` with `inner` type of join (perhaps, some other operations). However, it turned out the `merge` was slower. I don't see why.

Reply:
@devin-petersohn , I am sorry, I explained the behavior for `merge` with `inner` type of join a little bit incorrectly. Yes, we need to have the same axis (`axis=0`) to perform `merge` with `inner` type of join. I edited the previous comment, However, saying about the performance I still have a question why merge was slower.